liquidctl/driver/smbus.py

```diff
@@ -0,0 +1,275 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+if sys.platform == 'linux':
+
+    # smbus is an optional dependency
+    try:
+        from smbus import SMBus
+    except ModuleNotFoundError:
+        SMBus = None
+
+
+    class LinuxI2c(BaseBus):
+        """The Linux I²C (`/sys/bus/i2c`) bus."""
+
+        def find_devices(self, bus=None, usb_port=None, **kwargs):
+            """Find compatible SMBus devices."""
+
+            if usb_port:
+                # a usb_port filter implies an USB bus
+                return
+
+            if not SMBus:
+                _LOGGER.debug('skipping %s, smbus package not available',
+                              self.__class__.__name__)
+                return
+
+            devices = Path('/sys/bus/i2c/devices')
+            if not devices.exists():
+                _LOGGER.debug('skipping %s, %s not available',
+                              self.__class__.__name__, devices)
+                return
+
+            drivers = sorted(find_all_subclasses(SmbusDriver),
+                             key=lambda x: (x.__module__, x.__name__))
+
+            _LOGGER.debug('searching %s (%s)', self.__class__.__name__,
+                          ', '.join(map(lambda x: x.__name__, drivers)))
+
+            for i2c_dev in devices.iterdir():
+                if not i2c_dev.name.startswith('i2c-'):
+                    continue  # wont have a matching /dev/i2c-* file
+
+                i2c_bus = LinuxI2cBus(i2c_dev)
+
+                if bus and bus != i2c_bus.name:
+                    continue
+
+                _LOGGER.debug('found I²C bus %s', i2c_bus.name)
+                yield from i2c_bus.find_devices(drivers, **kwargs)
+
+
+    class LinuxI2cBus:
+        """A Linux I²C device, which is itself an I²C bus.
+
+        Should not be instantiated directly; use `LinuxI2c.find_devices`
+        instead.
+
+        This type mimics the `smbus.SMBus` read/write/close APIs.  However,
+        `open` does not take any parameters, and not all APIs are available.
+        """
+
+        # note: this is not a liquidctl BaseBus, as that would cause
+        # find_liquidctl_devices to try to directly instantiate it
+
+        def __init__(self, i2c_dev):
+            self._i2c_dev = i2c_dev
+            self._smbus = None
+
+            assert i2c_dev.name.startswith('i2c-')
+            self._number = int(i2c_dev.name[4:])
+
+        def find_devices(self, drivers, **kwargs):
+            """Probe drivers and find compatible devices in this bus."""
+            for drv in drivers:
+                yield from drv.probe(self, **kwargs)
+
+        def open(self):
+            """Open the I²C bus."""
+            if not self._smbus:
+                try:
+                    self._smbus = SMBus(self._number)
+                except FileNotFoundError:
+                    if Path('/sys/class/i2c-dev').exists():
+                        raise
+                    raise OSError('kernel module i2c-dev not loaded') from None
```

Comment:
why is this error `from None` I'm not all too familiar with that syntax. From what I Googled doesn't that hide the original exception?

Reply:
I also only found out about this yesterday, despite it being documented in the introduction of [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html).

TL;DR: when using `from None` the original exception is kept in `__context__`, but is otherwise hidden; this avoids the display of double tracebacks.

Reply:
Would that be something that should be refactored elsewhere in other modules?

---
I will test that out a little to see how it shows expresses itself to learn that.

Reply:
Yes, and we should also switch to custom exceptions when applicable.